add a Prolog solution (tested with SWI-Prolog)

gasche opened the pull request:
It is too hard to setup a Curry system (according to my experiments
with nqueens, KICS2 is very slow on search problems and PAKS requires
an existing Prolog setup, making it strictly harder to install than
a Prolog), so I'm adding a native Prolog version which we may use for
comparison instead.

Tested with SWI-prolog. The `prolog.sh script` gives an interface
compatible with `run_benchmark_time.rb`:

```
$ ruby ./run_benchmark_time.rb "sh prolog.sh"
sh prolog.sh
4: 89.0
5: 92.2
6: 91.4
7: 92.6
8: 115.4
9: 195.6
10: 637.4
11: 3086.8
12: 18689.4
```

gasche committed:
add a Prolog solution (tested with SWI-Prolog)

It is too hard to setup a Curry system (according to my experiments
with nqueens, KICS2 is very slow on search problems and PAKS requires
an existing Prolog setup, making it strictly harder to install than
a Prolog), so I'm adding a native Prolog version which we may use for
comparison instead.

Tested with SWI-prolog. The `prolog.sh script` gives an interface
compatible with `run_benchmark_time.rb`:

```
$ ruby ./run_benchmark_time.rb "sh prolog.sh"
sh prolog.sh
4: 89.0
5: 92.2
6: 91.4
7: 92.6
8: 115.4
9: 195.6
10: 637.4
11: 3086.8
12: 18689.4
```

jkoppel commented:
Welp; I thought I knew some basic Prolog, but I had trouble reading this. I think I'd need you to walk me through it.

So, to be clear, are you proposing removing Curry from the benchmarks? If you can run KICS2, that should be enough, because I've already found that PAKCS is slower.

Still, since this is much faster than either Curry implementation, I support including this in the benchmarks. We can still mention how both implementations timed out in informal testing.

----------------------------------

It's just one week till the deadline! What is your working plan from now till then? (What you plan to do and when.) When do you plan to switch to writing? I'd still like to get a chance to Skype you to make sure we're on the same page.

I have tomorrow open to work on research. I hope to also have next Wed-Sat available for research. Would you like me to do anything on this paper? Is there anything that I'd be significantly advantaged on doing? Else, my default will be to do some work for our OOPSLA submission next month.

Keep up the good work!

P.S.: You got the E-mail I forwarded from Ben Lippmeier, right?

gasche commented:
I added more comments to the Prolog code which in particular clarify the relation with the direct-style ML implementations. The code is mostly a direct translation in relational style (with an extra argument for the return value to most functions), except for what I initially called `range` that does not compute a list but introduces non-determinism, which I renamed into `choose_in_range` for clarity.

gasche commented:
Yes, I would be in favor of removing the Curry code from the paper, to simplify builds on my end. Prolog gives reasonable numbers on how a production-ready language with native non-determinism behaves on this program -- it's interesting that it is sensibly slower than the replay or thermometers-based implementations, which I think demonstrates that N-queens is a best-case scenario (lots of search but almost no pure computation with replay costs).

(I'll email about the other points :-)

jkoppel commented:
Okay. I understand the code now. So, the two uses of "count" on line 28 are different? That throws me.
